### DISCUSSION TITLE
ValueError: [E002] Can't find factory for 'relation_extractor' for language English (en)

### DISCUSSION THREAD

**Discussion by HarshadaPatil26** [explosion/spaCy#7378] [Help: Coding & Implementations] [answered]
I've come across the issue:
ValueError: [E002] Can't find factory for 'relation_extractor' for language English (en). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).

Available factories: attribute_ruler, tok2vec, merge_noun_chunks, merge_entities, merge_subtokens, token_splitter, parser, beam_parser, entity_linker, ner, beam_ner, entity_ruler, lemmatizer, tagger, morphologizer, senter, sentencizer, textcat, textcat_multilabel, en.lemmatizer

I am referring the link for realtion extractor:
https://github.com/explosion/projects/tree/v3/tutorials/rel_component

**Comment by svlandeg** [comment]
Hi! Can you provide a little more background as to which command exactly you're running? Are you running the example project as is, or are you trying to integrate the code in a custom code base?

**Reply by HarshadaPatil26** [reply]
I am running the example project as is.
After training model if i try to load model,
nlp=spacy.load(("/myproject/rel_component/training/model-best")
Then getting below error.
ValueError: [E002] Can't find factory for 'relation_extractor' for language English (en). This usually happens when spaCy calls nlp.create_pipe with a custom component name that's not registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator @Language.component (for function components) or @Language.factory (for class components).

Available factories: attribute_ruler, tok2vec, merge_noun_chunks, merge_entities, merge_subtokens, token_splitter, parser, beam_parser, entity_linker, ner, beam_ner, entity_ruler, lemmatizer, tagger, morphologizer, senter, sentencizer, textcat, textcat_multilabel, en.lemmatizer

**Reply by svlandeg** [reply, marked as answer]
Ah, that makes sense. The code that is loading the model also needs to know about the new, custom factories that you used, so you'll have to import the relevant functions:
```
# make the factory work
from rel_pipe import make_relation_extractor, score_relations

# make the config work
from rel_model import create_relation_model, create_classification_layer, create_instances, 
```

**Reply by HarshadaPatil26** [reply]
I have imported the above functions.. still it showing same error.

**Reply by pshrutiii** [reply]
I've also been trying to use rel_component tutorial project from git as is. Though the model can be trained, I'm also not able to use it 
`nlp = spacy.load("./training/model-best")`
the above command keeps breaking with the same error as in this discussion. I tried your suggestion and added one of the imports but it gave me a new set of errors.
![Screen Shot 2021-03-11 at 11 04 10 AM](https://user-images.githubusercontent.com/4038855/110840401-87625200-8259-11eb-8120-50a20a3b1508.png)
When you open rel_pipe.py, you'll see it has a factory method
![Screen Shot 2021-03-11 at 11 05 10 AM](https://user-images.githubusercontent.com/4038855/110840723-ed4ed980-8259-11eb-9c8b-1a7326a683de.png)
Not quite sure what i'm doing wrong. I would appreciate any help to load the model.

**Reply by svlandeg** [reply]
Hi @pshrutiii!

Your code imports `relation_extractor`, but that's not actually a function or variable defined in `rel_pipe.py`. It's the name given to the function `make_relation_extractor`, and this name should be loaded into the registry when you import the corresponding function. Can you try again with the original snippet I cited:
```
# make the factory work
from rel_pipe import make_relation_extractor, score_relations

# make the config work
from rel_model import create_relation_model, create_classification_layer, create_instances, 
```
and let me know whether it works, or what errors you're getting?

**Reply by pshrutiii** [reply]
For the registry to pick up, which file should i modify with these imports? The closest match to your imports suggestion was _custom_functions.py_ and _evaluate.py_ - Since only _custom_functions.py_ was missing one of the imports, i tweaked that and hit: `!python -m spacy project run all_gpu` again when i try to load:  `nlp = spacy.load("./training/model-best")` -- it gives me the same error
![Screen Shot 2021-03-11 at 2 15 51 PM](https://user-images.githubusercontent.com/4038855/110862410-4deb1000-8274-11eb-8b96-88125e73363d.png)

**Reply by svlandeg** [reply]
You don't have to modify any of the existing files, you need to add the import in the file where you're typing `nlp = spacy.load(...model-best)` - i.e. below the `import spacy` statement in the middle screen of the picture you posted.

**Reply by pshrutiii** [reply]
ahhhh.... yeah it worked!

**Reply by Erik-Sen** [reply]
Which package do i need to install to use the import of @svlandeg?

**Reply by svlandeg** [reply]
We provide this REL project as example code as part of a tutorial - you can find the code here: https://github.com/explosion/projects/tree/v3/tutorials/rel_component

And more information about spaCy projects here: https://spacy.io/usage/projects

But it's not a separate package as such. The implementation is relatively simple, and for realistic use-case you might want to build further on this code first.

**Reply by johnsonpeng5** [reply]
![image](https://user-images.githubusercontent.com/31191581/124686741-7da7ce80-de88-11eb-85b9-59dcf071f736.png)

I think I figured out pshrutili's problem; but I'm stuck on the next step. Here's the link to the [Colab I'm using ](https://colab.research.google.com/drive/1jB0FYBTObQswncucTtRsJMcFVtjNElKs?authuser=2#scrollTo=o0SlNl2Ze7lx)

I hate to be an annoying internet noob; but is it possible to have a reproducible colab to open up and use to figure out how it works?

**Reply by cos-nho** [reply]
@HarshadaPatil26 did you ever figure this out? I have run into the same issue
